mito-ai/src/Extensions/AiChat/ChatMessage/ChatMessage.tsx

```diff
@@ -36,20 +37,72 @@ const ChatMessage: React.FC<IChatMessageProps> = ({
     operatingSystem,
     setDisplayCodeDiff,
     acceptAICode,
-    rejectAICode
+    rejectAICode,
+    onUpdateMessage
 }): JSX.Element | null => {
+    const [isEditing, setIsEditing] = useState(false);
+    const [editedContent, setEditedContent] = useState(
+        // When editing, remove the code block from message.
+        // ChatHistoryManager will re-add the code block.
+        (message.content as string).replace(/```[\s\S]*?```/g, '').trim()
+    );
+
     if (message.role !== 'user' && message.role !== 'assistant') {
         // Filter out other types of messages, like system messages
         return null
     }
 
     const messageContentParts = splitStringWithCodeBlocks(message)
 
+    const handleEditClick = () => {
+        setIsEditing(true);
+    };
+
+    const handleSave = () => {
+        onUpdateMessage(messageIndex, editedContent);
+        setIsEditing(false);
+    };
+
+    const handleCancel = () => {
+        // When canceling, remove the code block from the message.
+        setEditedContent((message.content as string).replace(/```[\s\S]*?```/g, '').trim());
+        setIsEditing(false);
+    };
+
+    if (isEditing) {
+        return (
+            <div className={classNames(
+                "message",
+                { "message-user": message.role === 'user' },
+                { 'message-assistant': message.role === 'assistant' },
+            )}>
+                <textarea
+                    value={editedContent}
+                    onChange={(e) => setEditedContent(e.target.value)}
+                    onKeyDown={(e) => {
+                        // Enter key sends the message, but we still want to allow 
+                        // shift + enter to add a new line.
+                        if (e.key === 'Enter' && !e.shiftKey) {
```

Comment:
Let's add support so when the user presses `escape` it cancels editing if in editing mode and loses focus.